### ISSUE TITLE
Packet loss during persistance

### ISSUE THREAD

**Issue by time-trader** [aerosense-ai/data-gateway#6]
### What is the current behavior?

Strange (negative gap) packet loss after data persistence possibly due to buffer overflow:
Here is the output, with bytes in the buffer waiting printed out:

```
[2021-12-06 18:04:30,224 | INFO | data_gateway.packet_reader] 0.000000 bytes waiting
[2021-12-06 18:04:30,241 | INFO | data_gateway.packet_reader] 769.000000 bytes waiting
[2021-12-06 18:04:30,242 | INFO | data_gateway.packet_reader] 522.000000 bytes waiting
[2021-12-06 18:04:30,243 | INFO | data_gateway.packet_reader] 275.000000 bytes waiting
[2021-12-06 18:04:30,244 | INFO | data_gateway.packet_reader] 28.000000 bytes waiting
[2021-12-06 18:04:30,265 | INFO | data_gateway.packet_reader] 0.000000 bytes waiting
[2021-12-06 18:04:34,236 | INFO | data_gateway.persistence] Window 0 written to disk.
[2021-12-06 18:04:34,236 | INFO | data_gateway.persistence] Saving Baros_P data to csv file.
[2021-12-06 18:04:34,494 | INFO | data_gateway.persistence] Saving Baros_T data to csv file.
[2021-12-06 18:04:34,664 | INFO | data_gateway.persistence] Saving Constat data to csv file.
[2021-12-06 18:04:34,721 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,722 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,722 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,723 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,723 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,724 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,725 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,725 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,726 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,726 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,727 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,727 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,728 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,728 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,728 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,729 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,729 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,730 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,730 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,731 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,731 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,732 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,732 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,733 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,733 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,733 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,734 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,734 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,735 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,735 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,736 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,736 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,737 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,737 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,737 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,738 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,738 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,739 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,739 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,740 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,740 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,741 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,741 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,742 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,742 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,742 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,743 | INFO | data_gateway.packet_reader] 4095.000000 bytes waiting
[2021-12-06 18:04:34,743 | INFO | data_gateway.packet_reader] 3952.000000 bytes waiting
[2021-12-06 18:04:34,744 | INFO | data_gateway.packet_reader] 3705.000000 bytes waiting
[2021-12-06 18:04:34,744 | INFO | data_gateway.packet_reader] 3458.000000 bytes waiting
[2021-12-06 18:04:34,745 | INFO | data_gateway.packet_reader] 3211.000000 bytes waiting
[2021-12-06 18:04:34,745 | INFO | data_gateway.packet_reader] 2964.000000 bytes waiting
[2021-12-06 18:04:34,746 | INFO | data_gateway.packet_reader] 2717.000000 bytes waiting
[2021-12-06 18:04:34,746 | INFO | data_gateway.packet_reader] 2470.000000 bytes waiting
[2021-12-06 18:04:34,746 | INFO | data_gateway.packet_reader] 2223.000000 bytes waiting
[2021-12-06 18:04:34,747 | INFO | data_gateway.packet_reader] 1976.000000 bytes waiting
[2021-12-06 18:04:34,747 | INFO | data_gateway.packet_reader] 1729.000000 bytes waiting
[2021-12-06 18:04:34,748 | INFO | data_gateway.packet_reader] 1482.000000 bytes waiting
[2021-12-06 18:04:34,748 | INFO | data_gateway.packet_reader] 1235.000000 bytes waiting
[2021-12-06 18:04:34,749 | INFO | data_gateway.packet_reader] 988.000000 bytes waiting
[2021-12-06 18:04:34,749 | INFO | data_gateway.packet_reader] 741.000000 bytes waiting
[2021-12-06 18:04:34,749 | INFO | data_gateway.packet_reader] 494.000000 bytes waiting
[2021-12-06 18:04:34,750 | INFO | data_gateway.packet_reader] 247.000000 bytes waiting
[2021-12-06 18:04:34,750 | INFO | data_gateway.packet_reader] 1023.000000 bytes waiting
[2021-12-06 18:04:34,751 | INFO | data_gateway.packet_reader] 776.000000 bytes waiting
[2021-12-06 18:04:34,751 | INFO | data_gateway.packet_reader] 529.000000 bytes waiting
[2021-12-06 18:04:34,752 | INFO | data_gateway.packet_reader] 282.000000 bytes waiting
[2021-12-06 18:04:34,752 | INFO | data_gateway.packet_reader] 35.000000 bytes waiting
[2021-12-06 18:04:34,766 | INFO | data_gateway.packet_reader] 7.000000 bytes waiting
[2021-12-06 18:04:34,766 | WARNING | data_gateway.packet_reader] Lost Baros_P packet(s): -57952858.82019043 ms gap
[2021-12-06 18:04:34,766 | WARNING | data_gateway.packet_reader] Lost Baros_T packet(s): -57952858.82019043 ms gap
[2021-12-06 18:04:34,811 | INFO | data_gateway.packet_reader] 741.000000 bytes waiting
[2021-12-06 18:04:34,811 | WARNING | data_gateway.packet_reader] Lost Baros_P packet(s): 57949038.85437011 ms gap
[2021-12-06 18:04:34,811 | WARNING | data_gateway.packet_reader] Lost Baros_T packet(s): 57949038.85437011 ms gap
[2021-12-06 18:04:34,812 | INFO | data_gateway.packet_reader] 494.000000 bytes waiting
[2021-12-06 18:04:34,812 | INFO | data_gateway.packet_reader] 247.000000 bytes waiting
[2021-12-06 18:04:34,813 | INFO | data_gateway.packet_reader] 0.000000 bytes waiting
```

### Possible solutions:
@tommasopolonelli or @hmllr I would love your feedback on this. 

`set_buffer_size` is not an option in Linux, and on windows it is more of "recommendation" to the driver. 

- Create our own buffer in the gateway code and manage it.
- Put persistence operations on the separate thread from serial port reading
-  Use new and "experimental" [Packetizer class](https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.threaded.Packetizer) when handling packets from serial port

### Your environment

  - Library Version: `0.7.1`

**Comment by time-trader** [comment]
I ran the threaded version. (took out closing and reopening the port)
Here is the output of the console:
```
[2022-01-06 12:27:52,827 | INFO | MainThread | data_gateway.cli] Starting packet reader.
[2022-01-06 12:27:52,827 | INFO | MainThread | data_gateway.cli] Files will be uploaded to cloud storage at intervals of 60.0 seconds.
[2022-01-06 12:28:02,840 | INFO | MainThread | data_gateway.routine] Routine ran 'startDiffBaros' command.
[2022-01-06 12:28:18,197 | INFO | ParserThread | data_gateway.packet_reader] 
Exiting sleep

[2022-01-06 12:28:18,258 | INFO | ParserThread | data_gateway.packet_reader] Received first Diff_Baros packet
[2022-01-06 12:28:32,750 | INFO | ParserThread | data_gateway.packet_reader] Received first Constat packet
[2022-01-06 12:28:35,809 | WARNING | ParserThread | data_gateway.packet_reader] Possible packet loss. Constat sensor packet is timestamped 1960 ms later than expected
[2022-01-06 12:28:56,764 | INFO | ParserThread | data_gateway.persistence] Window 0 uploaded to cloud.
[2022-01-06 12:29:58,478 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:30:03,504 | INFO | ParserThread | data_gateway.persistence] Window 1 uploaded to cloud.
[2022-01-06 12:31:10,471 | INFO | ParserThread | data_gateway.persistence] Window 2 uploaded to cloud.
[2022-01-06 12:32:12,289 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,305 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,306 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,307 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,307 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,308 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,308 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:12,309 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:32:17,651 | INFO | ParserThread | data_gateway.persistence] Window 3 uploaded to cloud.
[2022-01-06 12:33:24,439 | INFO | ParserThread | data_gateway.persistence] Window 4 uploaded to cloud.
[2022-01-06 12:34:31,420 | INFO | ParserThread | data_gateway.persistence] Window 5 uploaded to cloud.
[2022-01-06 12:35:38,418 | INFO | ParserThread | data_gateway.persistence] Window 6 uploaded to cloud.
[2022-01-06 12:36:45,466 | INFO | ParserThread | data_gateway.persistence] Window 7 uploaded to cloud.
[2022-01-06 12:37:47,251 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:37:47,253 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:37:47,253 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:37:52,466 | INFO | ParserThread | data_gateway.persistence] Window 8 uploaded to cloud.
[2022-01-06 12:38:54,274 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:38:54,275 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:38:54,276 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:38:54,277 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:38:54,278 | WARNING | ReaderThread | data_gateway.packet_reader] Buffer is full. Some data may be lost.
[2022-01-06 12:38:59,375 | INFO | ParserThread | data_gateway.persistence] Window 9 uploaded to cloud.
[2022-01-06 12:40:06,422 | INFO | ParserThread | data_gateway.persistence] Window 10 uploaded to cloud
```
Good news: overflow does not happen on Baro sensors
Bad news: Buffer still gets full sometimes, and overflow still happens on Mics.

**Comment by time-trader** [comment]
Fixed with #30
